### ISSUE TITLE
Capitalised user meta key of Postcode results in deprecated notice and continual writing of billing_postcode

### ISSUE THREAD

**Issue by l3rady** [woocommerce/woocommerce#22534]
In our installation, we use the user meta key of postcode for other developed software. In the live environment, we came across the following error: over and over again
```
PHP message: The WC_Customer::set_postcode function is deprecated since version 3.0. Replace with WC_Customer::set_billing_postcode
```
We looked through the entire codebase and nowhere were we calling set_postcode. We assumed something was calling via a magic method or via a variable variable. We did a stacktrace and got:

```
1) 20190123103709/web/index.php(17): require('/var/www/www....')
2) 20190123103709/web/wp/wp-blog-header.php(13): require_once('/var/www/www....')
3) 20190123103709/web/wp/wp-load.php(42): require_once('/var/www/www....')
4 20190123103709/web/wp-config.php(63): require_once('/var/www/www....')
5) 20190123103709/web/wp/wp-settings.php(450): do_action('init')
6) 20190123103709/web/wp/wp-includes/plugin.php(453): WP_Hook->do_action(Array)
7) 20190123103709/web/wp/wp-includes/class-wp-hook.php(310): WP_Hook->apply_filters(NULL, Array)
8) 20190123103709/web/wp/wp-includes/class-wp-hook.php(286): WooCommerce->init('')
9) 20190123103709/web/app/plugins/woocommerce/includes/class-woocommerce.php(480): WC_Customer->__construct(30382, true)
10) 20190123103709/web/app/plugins/woocommerce/includes/class-wc-customer.php(104): WC_Data_Store->read(Object(WC_Customer))
11) 20190123103709/web/app/plugins/woocommerce/includes/class-wc-data-store.php(156): WC_Customer_Data_Store->read(Object(WC_Customer))
12) 20190123103709/web/app/plugins/woocommerce/includes/data-stores/class-wc-customer-data-store.php(163): WC_Data->set_props(Array)
13) 20190123103709/web/app/plugins/woocommerce/includes/abstracts/abstract-wc-data.php(622): WC_Legacy_Customer->set_postcode('SE21 8DW')
```

So on every page load a logged in user was triggering the error and it was WC_Customer_Data_Store->read that was triggering the set_postcode via set_props. Looking at the code there you deliberatley ignore meta `postcode`... urm then why are you not ignoring our postcode meta? Well turns out that half of our user have the user_meta_key as `Postcode`, notice the capital P. Your code doesn't account for casesensitivity. Because of that you were not ignoring postcode and then attempting to set the postcode on every page load...

As a temporary fix we have had to run the following query on our database:

```
update wp_usermeta set meta_key = 'postcode' where meta_key COLLATE utf8mb4_bin = 'Postcode';
```

I think this code you have:
```
		// Load meta but exclude deprecated props.
		$user_meta = array_diff_key(
			array_map( 'wc_flatten_meta_callback', get_user_meta( $customer_id ) ),
			array_flip( array( 'country', 'state', 'postcode', 'city', 'address', 'address_2', 'default', 'location' ) )
		);
```
should be made more robust to allow for case sensitivity or the callback function `wc_flatten_meta_callback` does it.

Without a fix the above creates a tonne of errors and a database write for logged in users for every page load. For us that is a major problem with 700K+ users.

**Comment by rrennick** [comment]
Can you provide your system status report? The line number in your debug log is in a comment in version 3.5.4

`13) 20190123103709/web/app/plugins/woocommerce/includes/abstracts/abstract-wc-data.php(622): WC_Legacy_Customer->set_postcode('SE21 8DW')`

https://github.com/woocommerce/woocommerce/blob/3.5.4/includes/abstracts/abstract-wc-data.php#L622

**Comment by l3rady** [comment]
Hi,

Sorry, I cannot as we are using an older version, I don't want to put anything on a public forum that could indicate which site it is until we migrate to the latest version, which is underway.

Saying that though the bug is still present in the latest version of woocommerce.

The issue is in the read function https://github.com/woocommerce/woocommerce/blob/3.5.4/includes/data-stores/class-wc-customer-data-store.php#L144

This is triggered on every page load for a logged in user. The code reads all of the user meta, filters out legacy meta keys and then uses set props to set the data back. Why the heck woocommerce writes back all the usermeta in a read function is beyond me, but it does.

The problem starts with this: https://github.com/woocommerce/woocommerce/blob/3.5.4/includes/data-stores/class-wc-customer-data-store.php#L155

It filters out the meta key `postcode` amongst others. The problem is that we have users that have the meta_key `Postcode` and so the code doesn't filter this out, because it is looking for `postcode`. This then gets sent to set props https://github.com/woocommerce/woocommerce/blob/3.5.4/includes/data-stores/class-wc-customer-data-store.php#L160

Then as it is looping over the meta keys the following is called: https://github.com/woocommerce/woocommerce/blob/3.5.4/includes/abstracts/abstract-wc-data.php#L645 This results in `$this->set_Postcode('foo')` which gets picked up as `$this->set_postcode('foo')` in: https://github.com/woocommerce/woocommerce/blob/3.5.4/includes/legacy/class-wc-legacy-customer.php#L252

This then triggers the wc_deprecated_function and results in setting the billing postcode.

Hope that clears things up.

**Comment by rrennick** [comment]
> Saying that though the bug is still present in the latest version of woocommerce.

The part of that line that is causing the warning is `WC_Legacy_Customer->set_postcode('SE21 8DW')`.

If you search the 3.5.4 version code you will only find `set_postcode` twice (the function declaration & the deprecated notice):

```
.//includes/legacy/class-wc-legacy-customer.php:252:	function set_postcode( $postcode ) {
.//includes/legacy/class-wc-legacy-customer.php:253:		wc_deprecated_function( 'WC_Customer::set_postcode', '3.0', 'WC_Customer::set_billing_postcode' );
```

**Comment by rrennick** [comment]
> Your code doesn't account for casesensitivity.

In PHP, strings are case sensitive while class names & function names are not. 'Postcode' and 'postcode' are unequal strings. This is resulting in the 'Postcode' meta key not being found but set_Postcode matching the deprecated function. 

The problem with switching from case sensitive to case insensitive is the potential to create cross pollination with meta that already exists in the DB that WooCommerce didn't recognize due to uppercase letters in the meta key.

**Comment by l3rady** [comment]
@rrennick in both latest version of woocommerce and the version I am using, the only occurrence of `set_postcode()` is in includes/legacy/class-wc-legacy-customer.php.

In woocommerce core code, it calls this function by a variable variable so any code search for usage of that function shows up nothing. Still, woocomerce in latest version, can call this legacy function by the problem code in `WC_Customer_Data_Store->read`

> The problem with switching from case sensitive to case insensitive is the potential to create cross pollination with meta that already exists in the DB that WooCommerce didn't recognize due to uppercase letters in the meta key.

This is not a cross-pollination issue here. The problem is this code:

```
$user_meta = array_diff_key(
			array_map( 'wc_flatten_meta_callback', get_user_meta( $customer_id ) ),
			array_flip( array( 'country', 'state', 'postcode', 'city', 'address', 'address_2', 'default', 'location' ) )
		);
```

is trying to ignore meta that would result in calling deprecated functions.

So let's say I have a user with this meta field: `foo` with the value of `bar`, the code in the read function will then call `set_props(['foo' => 'bar'])` which in turn will look for the method `set_foo('bar')` and because this method doesn't exist it ignores calling it. Thus woocommerce ignores meta provided by other plugins etc. If you now change that meta key to `Foo` the same thing functions as it should. Capitalization isn't an issue here.

Now woocommerce is interested in its own user meta it wants to deal with such as `set_billing_first_name` and has a method for that. So if woocommerce comes across that property (`billing_first_name`) it will send it off to the `set_billing_first_name('foo')` method. Now in the database that meta key was `BILLING_FIRST_NAME` you would still call `set_billing_first_name('foo')`. Again capitalisation, not an issue here.

Now at some point in the past you no longer deal with the following user meta keys `'country', 'state', 'postcode', 'city', 'address', 'address_2', 'default', 'location'` but you still have set methods for each of them in the legacy file. You keep the legacy set methods around just in case other plugins/code still call on them but internally code was written to ignore these meta keys and stop the deprecated functions from being called. But the code assumes the meta keys in the database are all lower case. If the meta_key `Postcode` is given the check misses it and passes this meta key to your built-in deprecated function.

Changing the code here to take into account the case-sensitivity of the keys will only stop existing keys from triggering deprecated functions that you want to avoid in the first place?

I think the code should be changed to the following:

```
        // Load meta but exclude deprecated props.
        $user_meta = array_map('wc_flatten_meta_callback', get_user_meta($customer_id));
        $ignored_user_meta = array_intersect_key(
            array_change_key_case($user_meta),
            array_flip(array('country', 'state', 'postcode', 'city', 'address', 'address_2', 'default', 'location'))
        );
        $user_meta = array_diff_key($user_meta, $ignored_user_meta);
```

This code makes sure that it only lower cases the keys you are trying to ignore due to legacy and then removes the found keys from `$user_meta`

**Comment by rrennick** [comment]
> This is not a cross-pollination issue here.

It wouldn't be a cross pollination issue for you. I was referring to all of the other WC installs that might have meta_key with uppercase letters in them. WC currently ignores those meta rows.

**Comment by l3rady** [comment]
> WC currently ignores those meta rows.

Currently, it isn't because the code is bugged and I think it was an oversight when implemented that if the meta keys are not all lower case then the legacy code is triggered when it shouldn't.

> It wouldn't be a cross pollination issue for you.

I fail to see how it is a cross-pollination issue for anyone? Could you maybe give me a scenario where the change in that bit of code to ignore all case meta keys would cause an issue?

**Comment by mikejolley** [comment]
This was fixed in 3.5.4.

https://github.com/woocommerce/woocommerce/commit/3943ec639592d42512d7b7445383859356ba6cd6

This converts `Postcode` to `postcode` then excludes. I was unable to replicate the issue on master 

> Sorry, I cannot as we are using an older version, I don't want to put anything on a public forum that could indicate which site it is until we migrate to the latest version, which is underway.

Please provide this in future otherwise it can lead to unneccessary back and forth. You don't need to share your web address if sensistive but unless we're sure of version it's hard to confirm or replicate issues :)

Thanks
